crates/nri-resctrl-plugin/src/lib.rs

```
@@ -332,6 +341,196 @@ impl<P: FsProvider> ResctrlPlugin<P> {
+            match st.pods.get(pod_uid) {
+                Some(ps) => matches!(ps.group_state, ResctrlGroupState::Failed),
+                None => {
+                    return Err(resctrl::Error::Io {
+                        path: PathBuf::from("<pod>"),
+                        source: io::Error::from_raw_os_error(libc::ENOENT),
+                    })
+                }
+            }
+        };
+
+        if !should_try {
+            // Either does not exist or already Exists; return current state
+            let st = self.state.lock().unwrap();
+            let current = st.pods.get(pod_uid).unwrap().group_state.clone();
+            return Ok(current);
+        }
+
+        // Drop lock while performing filesystem operation
+        let res = self.resctrl.create_group(pod_uid);
+        match res {
+            Ok(path) => {
+                let mut st = self.state.lock().unwrap();
+                // Re-check and update if still Failed
+                if let Some(ps) = st.pods.get_mut(pod_uid) {
+                    if matches!(ps.group_state, ResctrlGroupState::Failed) {
+                        ps.group_state = ResctrlGroupState::Exists(path.clone());
+                        let snapshot = ps.clone();
+                        drop(st);
+                        self.emit_pod_add_or_update(pod_uid, &snapshot);
+                        return Ok(ResctrlGroupState::Exists(path));
+                    } else if let ResctrlGroupState::Exists(p) = &ps.group_state {
+                        return Ok(ResctrlGroupState::Exists(p.clone()));
+                    }
+                }
+                // Pod disappeared concurrently
+                Err(resctrl::Error::Io {
+                    path: std::path::PathBuf::from("<pod>"),
+                    source: std::io::Error::from_raw_os_error(libc::ENOENT),
+                })
+            }
+            Err(resctrl::Error::Capacity { .. }) => Err(resctrl::Error::Capacity {
```

let's just return the error as is, overriding its source doesn't make sense -- user should see the actual source. So we can merge this with the Err(e) => Err(e) branch.